panther_manager/test/test_safety_behavior_tree.cpp

```diff
@@ -0,0 +1,422 @@
+    "hardware/aux_power_enable",
+    std::bind(&TestSafetyBehaviorTree::AUXPowerServerCB, this, _1, _2));
+  e_stop_trigger_server_ = safety_manager_node_->create_service<TriggerSrv>(
+    "hardware/e_stop_trigger", std::bind(&TestSafetyBehaviorTree::EStopTriggerSrvCB, this, _1, _2));
+}
+
+TestSafetyBehaviorTree::~TestSafetyBehaviorTree() { rclcpp::shutdown(); }
+
+void TestSafetyBehaviorTree::SpinWhileRunning()
+{
+  test_behavior_tree_utils::SpinWhileRunning(
+    safety_manager_node_, [&]() { return safety_manager_node_->GetSafetyTreeStatus(); });
+}
+
+void TestSafetyBehaviorTree::PublishDefaultStates()
+{
+  PublishEStop(false);
+  PublishBatteryState(
+    BatteryStateMsg::POWER_SUPPLY_STATUS_DISCHARGING, BatteryStateMsg::POWER_SUPPLY_HEALTH_GOOD,
+    0.0);
+  PublishIOState(false, true);
+  PublishDriverState(0.0);
+  PublishSystemStatus(0.0);
+}
+
+void TestSafetyBehaviorTree::PublishEStop(const bool data)
+{
+  BoolMsg msg;
+  msg.data = data;
+  e_stop_pub_->publish(msg);
+}
+
+void TestSafetyBehaviorTree::PublishBatteryState(
+  const std::uint8_t status, const std::uint8_t health, const float temperature)
+{
+  BatteryStateMsg msg;
+  msg.power_supply_status = status;
+  msg.power_supply_health = health;
+  msg.temperature = temperature;
+  battery_state_pub_->publish(msg);
+}
+
+void TestSafetyBehaviorTree::PublishIOState(const bool fan, const bool aux_power)
+{
+  IOStateMsg msg;
+  msg.fan = fan;
+  msg.aux_power = aux_power;
+  io_state_pub_->publish(msg);
+}
+
+void TestSafetyBehaviorTree::PublishDriverState(const float temperature)
+{
+  DriverStateMsg msg;
+  msg.front.temperature = temperature;
+  msg.rear.temperature = temperature;
+  driver_state_pub_->publish(msg);
+}
+
+void TestSafetyBehaviorTree::PublishSystemStatus(const float cpu_temp)
+{
+  SystemStatusMsg msg;
+  msg.cpu_temp = cpu_temp;
+  system_status_pub_->publish(msg);
+}
+
+void TestSafetyBehaviorTree::FanServerCB(
+  const SetBoolSrv::Request::SharedPtr & request, SetBoolSrv::Response::SharedPtr response)
+{
+  fan_state_ = request->data;
+  PublishIOState(fan_state_, aux_power_state_);
+  response->success = true;
+}
+
+void TestSafetyBehaviorTree::AUXPowerServerCB(
+  const SetBoolSrv::Request::SharedPtr & request, SetBoolSrv::Response::SharedPtr response)
+{
+  aux_power_state_ = request->data;
+  PublishIOState(fan_state_, aux_power_state_);
+  response->success = true;
+}
+
+void TestSafetyBehaviorTree::EStopTriggerSrvCB(
+  const TriggerSrv::Request::SharedPtr & /*request*/, TriggerSrv::Response::SharedPtr response)
+{
+  e_stop_triggered_ = true;
+  response->success = true;
+}
+
+TEST_F(TestSafetyBehaviorTree, TurnOnFanAtStartup)
+{
+  safety_manager_node_->Initialize();
+  PublishDefaultStates();
+
+  SpinWhileRunning();
+  EXPECT_TRUE(fan_state_);
+
+  // wait for automatic turn off of the fan
+  std::this_thread::sleep_for(
+    std::chrono::milliseconds(static_cast<unsigned>(kFanTurnOffTimeout * 1000)));
```

Comment:
So much sleep is making unit testing enormously slow. Can we set the timeout to 0 or somehow speed them up?

Reply:
This is a bit more complicated because lack of this timeout can cause other test to fail. Here is how removing timeout can cause this:
Publish default states and spin -> tree calls TurOnFaAtStartup and returns SUCCESS
Publish some new info, then spin. -> tree timer is executed in callback first. Tree ticks first, topic callbacks are not executed yet. Because timeout is 0 it calls TurOffFan and returns SUCCESS. Now topic callbacks are executed. `SpinWhileRunning()` checks tree state which is SUCCESS and exits. Tree was never executed with updated values from topic callback and test fails.
If there is small timeout then `TurOffFan` wont be called during this process and tree will return skipped in first run and then will tick 2nd time when topic callbacks were already called.

The solution I can see is do additional spin after new message is published and then calling `SpinWhileRunning`. What do you think?

Reply:
A second spin is a better option than a timeout. 

The best option is to omit the whole publishing aspect. We could add a new method responsible for setting the variables directly in the blackboard and spin once the tree to read them and make decisions.
- The ROS part is already tested in `test_<name>_behavior_tree.cpp`
- We stay focused on the BT logic - simple and lightweight
- No wait is needed - much faster unit testing